### PR TITLE
Bug fix issue 30351

### LINKED ISSUE #30351: IndentedTextWriter Async methods don't indent properly
[System.CodeDom.Compiler.IndentedTextWriter](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/CodeDom/Compiler/IndentedTextWriter.cs) wrapped around a `StringWriter` doesn't indent properly when using the `WriteLineAsync` methods.

- First write is not indented
- Writes following empty lines (i.e. from `await WriteLineAsync();`) are not indented
- Lines following lines with content are indented correctly.

Gist: https://gist.github.com/markrendle/01b8b70d2b7c4f023bf196e96562c14a

N.B.: I know there's no point using `*Async` on a `StringWriter`, but the code in question is used with both `StringWriter` and `StreamWriter`.

N.B. 2: I'd just submit a PR to fix this, but I don't know if that would constitute a breaking change.

Using .NET Core 3.0 Preview 7 win-x64.

### PR DESCRIPTION
**Pull request by NewellClark**
This fixes [issue 30351](https://github.com/dotnet/runtime/issues/30351). I noticed there weren't any unit tests for System.CodeDom.IndentedTextWriter, so I went ahead and added those as well.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs [hunk 1]
```diff
@@ -141,6 +159,36 @@ public override void Write(string format, params object?[] arg)
             _writer.Write(format, arg);
         }
 
+        public override async Task WriteAsync(char value)
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:162]: I believe all of this new public APIs also need to be added in ref https://github.com/dotnet/runtime/blob/8d7e074fe88d05a66f70e1717c1bea490cd41d2e/src/libraries/System.Runtime/ref/System.Runtime.cs#L5148-L5190

File: src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj [hunk 2]
```diff
@@ -98,18 +99,12 @@
...
-    <Compile Include="$(CommonTestPath)System\Collections\ICollection.NonGeneric.Tests.cs"
-             Link="Common\System\Collections\ICollection.NonGeneric.Tests.cs" />
-    <Compile Include="$(CommonTestPath)System\Collections\IEnumerable.NonGeneric.Tests.cs"
-             Link="Common\System\Collections\IEnumerable.NonGeneric.Tests.cs" />
-    <Compile Include="$(CommonTestPath)System\Collections\TestBase.NonGeneric.cs"
-             Link="Common\System\Collections\TestBase.NonGeneric.cs" />
-    <Compile Include="$(CommonTestPath)System\IO\TempFile.cs"
-             Link="Common\System\IO\TempFile.cs" />
-    <Compile Include="$(CommonTestPath)System\Runtime\Serialization\Formatters\BinaryFormatterHelpers.cs"
-             Link="Common\System\Runtime\Serialization\Formatters\BinaryFormatterHelpers.cs" />
```
Comment (buyaa-n) [inline src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj:112]: Could you revert unwanted/unrelated changes, it would be cleaner for review

File: src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+        public async Task WriteLineAsync_WithIndents_FirstLine_IsNotIndented(Func<IndentedTextWriter, Task> callWriteLineAsync, string expected)
+        {
+            var indicator = new IndicatingTextWriter();
+            var itw = new IndentedTextWriter(indicator, TabString);
+            itw.NewLine = NewLine;
+            itw.Indent = 1;
+            await callWriteLineAsync(itw);
+
+            Assert.Equal(nameof(IndentedTextWriter.WriteLineAsync), indicator.LastCalledMethod);
+            Assert.Equal($"{expected}", indicator.GetStringBuilder().ToString());
```
Comment (buyaa-n) [inline src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs:470]:
```suggestion
            Assert.Equal(expected, indicator.GetStringBuilder().ToString());
```

File: src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs (hunk 1 above)
Comment (NewellClark) [reply src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:162]: Oh okay. I wasn't sure if that was necessary because they're virtual methods defined on a base class, which already has the refs. I'll go ahead and add them.

File: src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj (hunk 2 above)
Comment (NewellClark) [reply src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj]: My apologies. Not sure how that got changed. I guess VS changed them? I've reverted them.

File: src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+            var itw = new IndentedTextWriter(indicator, TabString);
+            itw.NewLine = NewLine;
+            itw.Indent = 1;
+
+            const string prefix = "prefix";
+            await itw.WriteLineAsync(prefix);
+            await callWriteLineAsync(itw);
+
+            Assert.Equal(nameof(IndentedTextWriter.WriteLineAsync), indicator.LastCalledMethod);
+            Assert.Equal($"{prefix}{NewLine}{TabString}{expected}", indicator.GetStringBuilder().ToString());
```
Comment (buyaa-n) [inline src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs:488]: Could we add a test testing the original scenario, that shows that indentation is added after calling `IndentedTextWriter.WriteLineAsync();`, something like:
```csharp
        [Fact]
        public async Task TestIdentationAfterParameterlessWriteLineAsync()
        {
            var asyncSw = new StringWriter();
            var asyncTw = new IndentedTextWriter(asyncSw);

            asyncTw.Indent++;
            await asyncTw.WriteLineAsync("Foo");
            await asyncTw.WriteLineAsync("Bar");
            await asyncTw.WriteLineAsync();
            await asyncTw.WriteLineAsync("Quux");
            await asyncTw.WriteLineAsync("Wibble");

            var sw = new StringWriter();
            var tw = new IndentedTextWriter(sw);

            tw.Indent++;
            tw.WriteLine("Foo");
            tw.WriteLine("Bar");
            tw.WriteLine();
            tw.WriteLine("Quux");
            tw.WriteLine("Wibble");

            Assert.Equal(sw.GetStringBuilder().ToString(), asyncSw.GetStringBuilder().ToString());
        }
```

File: src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj]: No problem, thanks

File: src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs (hunk 1 above)
Comment (NewellClark) [reply src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:162]: Okay, I've added them.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 5]
```diff
@@ -5156,7 +5156,9 @@ public partial class IndentedTextWriter : System.IO.TextWriter
         [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
         public override string NewLine { get { throw null; } set { } }
         public override void Close() { }
+        public override ValueTask DisposeAsync() => throw null;
```
Comment (buyaa-n) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:5159]: Oops, types need to have fully qualified name in ref
```suggestion
        public override System.Threading.Tasks.ValueTask DisposeAsync() => throw null;
```

Comment (NewellClark) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:5159]: Sorry about that. These "ref" projects are uncharted territory for me. I'll fix them and add the new unit tests.

Comment (buyaa-n) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:5159]: No problem at all, thank you for working on it

File: src/libraries/System.Runtime/System.Runtime.sln [hunk 6]
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.30804.86
+MinimumVisualStudioVersion = 10.0.40219.1
```
Comment (buyaa-n) [inline src/libraries/System.Runtime/System.Runtime.sln:5]: Oops sorry, again VS changed this, please revert

Comment (NewellClark) [reply src/libraries/System.Runtime/System.Runtime.sln]: Reverted.

**Comment by buyaa-n** [comment]
Thank you @NewellClark, the PR looks good to me, i have 2 questions for @stephentoub and @carlossanlop 
1. Do we want to add the APIs into ref? i don't see any reason not to except it is not gone through API review
2. I am not sure if the tests should have added in System.IO, i am OK having it in `CodeDom` though, just saw there is existing test for `IndentedTextWriter` in System.IO

**Comment by NewellClark** [comment]
> Thank you @NewellClark, the PR looks good to me, i have 2 questions for @stephentoub and @carlossanlop
> 
> 1. Do we want to add the APIs into ref? i don't see any reason not to except it is not gone through API review
> 2. I am not sure if the tests should have added in System.IO, i am OK having it in `CodeDom` though, just saw there is existing test for `IndentedTextWriter` in System.IO

I have just a few comments:
1) While we're discussing APIs, we may want to consider adding `protected virtual Task OutputTabsAsync()`, as well as `public Task WriteLineNoTabsAsync(string? s)`, to match the existing synchronous APIs that are specific to `IndentedTextWriter`. For example, if someone subclasses and overrides `OutputTabs()`, they should really also be overriding `OutputTabsAsync()`, otherwise they'll get the same bug in their subclass that we just fixed in `IndentedTextWriter`. 
2) Oops, My bad. I didn't see the existing tests in `System.IO`.

**Comment by buyaa-n** [comment]
> I can submit an API review if you'd like.

Sorry for late response @NewellClark, I was in doubt if its need to go through API review, in general all APIs added to ref should, but it is just adding missing overloads not adding anything new. I talk with @stephentoub offline and confirmed we don't need API review here

> While we're discussing APIs, we may want to consider adding `protected virtual Task OutputTabsAsync()`, as well as `public Task WriteLineNoTabsAsync(string? s)`, to match the existing synchronous APIs that are specific to `IndentedTextWrite`. For example, if someone subclasses and overrides `OutputTabs()`, they should really also be overriding `OutputTabsAsync()`, otherwise they'll get the same bug in their subclass that we just fixed in `IndentedTextWriter`.

That is good point, those are also looks like missing implementation, as synchronous versions already in the ref https://github.com/dotnet/runtime/blob/167c598c71b7f77fe8bf5db2bb113c2565d5c84d/src/libraries/System.Runtime/ref/System.Runtime.cs#L5160 I agree we should add Async versions too

> Oops, My bad. I didn't see the existing tests in System.IO.

Yeah it is confusing, from namespace i would think it is related to `CodeDome`, i guess it is added to `IO` because it is a writer, we might want to move tests to `IO` @carlossanlop what you think?

**Comment by NewellClark** [comment]
@buyaa-n Sounds good. I'll move the tests to `System.IO` and add the async versions of existing methods previously discussed.

**Review by buyaa-n — APPROVED** [review]
Thank you very much @NewellClark, looks good to me, lets wait for System.IO area owner @carlossanlop's review/approval

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 7]
```diff
@@ -5156,8 +5156,11 @@ public partial class IndentedTextWriter : System.IO.TextWriter
         [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
         public override string NewLine { get { throw null; } set { } }
         public override void Close() { }
+        public override System.Threading.Tasks.ValueTask DisposeAsync() => throw null;
```
Comment (carlossanlop) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:5159]: Here and elsewhere: I'm not sure if using `=>` would cause trouble, so let's stick to safety and use the same pattern used in all our refs (traditional parenthesis):
```suggestion
        public override System.Threading.Tasks.ValueTask DisposeAsync() { throw null; }
```

File: src/libraries/System.IO/tests/IndentedTextWriter.cs [hunk 8]
```diff
@@ -178,5 +465,330 @@ public static async Task Writes_ProducesExpectedOutput(string newline)
...
+                yield return CreateParameters(x => x.WriteLineAsync('c'), $"c{NewLine}");
+                yield return CreateParameters(x => x.WriteLineAsync("Hello World".ToCharArray(), 6, 5), $"World{NewLine}");
+                yield return CreateParameters(x => x.WriteLineAsync("Hello World"), $"Hello World{NewLine}");
+            }
+        }
+
+        private const string TabString = "   ";
+        private const string NewLine = "\n";
+
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
```
Comment (carlossanlop) [inline src/libraries/System.IO/tests/IndentedTextWriter.cs:562]: Did you have to add this because the CI failing in .NET Framework? If you didn't see any failures, maybe the attribute is not needed.

**Review by carlossanlop — CHANGES_REQUESTED** [review]
Looks good. I left a few suggestions for you to consider. Please add documentation so it can be approved.

File: src/libraries/System.IO/tests/IndentedTextWriter.cs (hunk 8 above)
Comment (NewellClark) [reply src/libraries/System.IO/tests/IndentedTextWriter.cs:562]: I just ran the tests without it and they passed. Not sure why I added them, it was a couple weeks ago. I'll remove them.

File: src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs [hunk 9]
```diff
@@ -49,8 +51,12 @@ public int Indent
 
         public override void Close() => _writer.Close();
 
+        public override ValueTask DisposeAsync() => _writer.DisposeAsync();
```
[the remark this answers is not in the document]
Comment (NewellClark) [reply src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:55]: I just ran the tests again and they passed without it, so I've gone ahead and removed it.

File: src/libraries/System.Runtime/ref/System.Runtime.cs (hunk 7 above)
Comment (NewellClark) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: Fixed

File: src/libraries/System.IO/tests/IndentedTextWriter.cs (hunk 8 above)
Comment (buyaa-n) [reply src/libraries/System.IO/tests/IndentedTextWriter.cs:562]: Well, CodeDome test project had .Net framework target `net461`, i guess it was failing before moving tests to System.IO https://github.com/dotnet/runtime/blob/32d664d10ce92f6535606b4ca658d0e9cfd754a4/src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj#L4

now Syste.IO doesn't have .Net framework target

**Comment by NewellClark** [comment]
@carlossanlop I've implemented your suggested changes.

**Review by carlossanlop — APPROVED** [review]

**Comment by buyaa-n** [comment]
Thank you @NewellClark, we can merge it now
